File: 3154-maximum-value-of-an-ordered-triplet-i/3154-maximum-value-of-an-ordered-triplet-i.go
func maximumTripletValue(nums []int) int64 {
    n := len(nums)
    var ans int64 = 0
    var dmax int64 = 0
    var imax int64 = 0
    for i := 0; i < n; i++ {
        a := dmax * int64(nums[i])
        d := imax - int64(nums[i])
        if a > ans {
            ans = a
        }
        if d > dmax {
            dmax = d
        }
        if int64(nums[i]) > imax {
            imax = int64(nums[i])
        }
    }
    return ans
}

File: 3154-maximum-value-of-an-ordered-triplet-i/3154-maximum-value-of-an-ordered-triplet-i.c
typedef long long ll;
long long maximumTripletValue(int* nums, int numsSize) {
    ll ans = 0, dmax = 0, imax = 0;
    for (int i = 0; i < numsSize; ++i) {
        ans = fmax(ans, dmax * nums[i]);
        dmax = fmax(dmax, imax - nums[i]);
        imax = fmax(imax, nums[i]);
    }
    return ans;
}